lib/codegen/src/licm.rs

```diff
@@ -111,21 +118,28 @@ fn has_pre_header(
     header: Ebb,
 ) -> Option<(Ebb, Inst)> {
     let mut result = None;
-    let mut found = false;
     for BasicBlock {
         ebb: pred_ebb,
         inst: last_inst,
     } in cfg.pred_iter(header)
     {
         // We only count normal edges (not the back edges)
         if !domtree.dominates(header, last_inst, layout) {
-            if found {
+            if result.is_some() {
                 // We have already found one, there are more than one
                 return None;
-            } else {
-                result = Some((pred_ebb, last_inst));
-                found = true;
             }
+            if last_inst != layout.last_inst(pred_ebb).unwrap() {
```

Comment:
Pre-existing: I think `last_inst` is a bad name (I was wondering why the last instruction wouldn't be equal to the last instruction!); maybe we could rename it `branch_inst`, or `pred_branch_inst` to make it clear that it's the actual branch instruction that jumps into the header?

Also, isn't this test redundant with the one checking the number of successors below?